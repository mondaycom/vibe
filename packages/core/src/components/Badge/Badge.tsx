import React, { forwardRef, useRef } from "react";
import cx from "classnames";
import { camelCase } from "lodash-es";
import useMergeRef from "../../hooks/useMergeRef";
import VibeComponentProps from "../../types/VibeComponentProps";
import VibeComponent from "../../types/VibeComponent";
import { getTestId } from "../../tests/test-ids-utils";
import { ComponentDefaultTestId } from "../../tests/constants";
import {
  BadgeAlignments as BadgeAlignmentsEnum,
  BadgeAnchor as BadgeAnchorEnum,
  BadgeType as BadgeTypeEnum
} from "./BadgeConstants";
import { BadgeAlignments, BadgeAnchor, BadgeType } from "./Badge.types";
import Indicator, { IndicatorProps } from "./Indicator/Indicator";
import Counter, { CounterProps } from "../Counter/Counter";
import { getStyle } from "../../helpers/typesciptCssModulesHelper";
import { IndicatorColor } from "./Indicator/IndicatorConstants";
import { CounterColor } from "../Counter/CounterConstants";
import styles from "./Badge.module.scss";

export interface BadgeBaseProps extends VibeComponentProps {
  /**
   * The position of the badge relative to its parent.
   */
  anchor?: BadgeAnchor;
  /**
   * The alignment style of the badge.
   */
  alignment?: BadgeAlignments;
  /**
   * The content the badge is attached to.
   */
  children: React.ReactNode;
}

interface CounterBadgeProps extends CounterProps {
  /**
   * The type of badge, set to `"counter"` for numeric values.
   */
  type: Extract<BadgeType, "counter">;
}

interface IndicatorBadgeProps extends IndicatorProps {
  /**
   * The type of badge, set to `"indicator"` for a simple dot.
   */
  type?: Extract<BadgeType, "indicator">;
}

export type BadgeProps = BadgeBaseProps & (CounterBadgeProps | IndicatorBadgeProps);

const Badge: VibeComponent<BadgeProps> & {
  types?: typeof BadgeTypeEnum;
  alignments?: typeof BadgeAlignmentsEnum;
  anchors?: typeof BadgeAnchorEnum;
} = forwardRef(
  (
    {
      type = "indicator",
      anchor = "top-end",
      alignment = "rectangular",
      className,
      id,
      "data-testid": dataTestId,
      children,
      ...badgeProps
    }: BadgeProps,
    ref
  ) => {
    const componentRef = useRef(null);
    const mergedRef = useMergeRef(ref, componentRef);
    const badgeClassNames = cx(
      styles.badge,
      getStyle(styles, camelCase(anchor as unknown as string)),
      getStyle(styles, alignment)
    );

    const color = badgeProps.color || type === "indicator" ? "notification" : "negative";

    return (
      <div ref={mergedRef} className={cx(styles.badgeWrapper, className)} id={id}>
        {children}
        <div className={badgeClassNames} data-testid={dataTestId || getTestId(ComponentDefaultTestId.BADGE, id)}>
          {type === "indicator" ? (
            <Indicator color={color as IndicatorColor} {...(badgeProps as IndicatorBadgeProps)} />
          ) : (
            type === "counter" && <Counter color={color as CounterColor} {...(badgeProps as CounterBadgeProps)} />
          )}
        </div>
      </div>
    );
  }
);

Badge.types = BadgeTypeEnum;
Badge.alignments = BadgeAlignmentsEnum;
Badge.anchors = BadgeAnchorEnum;

export default Badge;
