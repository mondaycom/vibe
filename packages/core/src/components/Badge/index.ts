export { default as Badge, type BadgeProps } from "./Badge";

export * from "./Badge.types";
