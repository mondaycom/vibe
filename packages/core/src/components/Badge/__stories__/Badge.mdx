import Badge from "../Badge";
import { Meta } from "@storybook/blocks";
import person from "./assets/person.png";
import { WhatsNew } from "@vibe/icons";
import Link from "../../Link/Link";
import Button from "../../Button/Button";
import Avatar from "../../Avatar/Avatar";
import { BUTTON, COUNTER, LINK } from "../../../storybook/components/related-components/component-description-map";
import Indicator from "../Indicator/Indicator";
import * as BadgeStories from "./Badge.stories";

<Meta of={BadgeStories} />

# Badge

- [Overview](#overview)
- [Props](#props)
- [Variants](#variants)
- [Do’s and don’ts](#dos-and-donts)
- [Use cases and examples](#use-cases-and-examples)
- [Related components](#related-components)
- [Feedback](#feedback)

## Overview

Badge component is responsible for layout an indicator/counter on a child component

<Canvas of={BadgeStories.Overview} />

### Import

```js
import { Badge } from "@vibe/core";
```

## Props

<PropsTable />

## Variants

### States

Badge can be of type Indicator or type Counter

<Canvas of={BadgeStories.States} />

### Button

When using Badge on a Button element, use alignment of RECTANGULAR in order to attach it to the element

<Canvas of={BadgeStories.ButtonStory} />

### Avatar

When using Badge on an Avatar element, use alignment of CIRCULAR in order to attach it to the element

<Canvas of={BadgeStories.AvatarStory} />

### Inline elements

When using Badge on an inline element, use alignment of OUTSIDE in order to attach it to the element

<Canvas of={BadgeStories.InlineElements} />

## Do’s and Don’ts

<ComponentRules
  rules={[
    {
      positive: {
        component: (
          <Badge alignment="outside">
            <Link text="Read more" />
          </Badge>
        ),
        description: "When using the badge on an inline component, apply OUTSIDE alignment to it"
      },
      negative: {
        component: (
          <Badge>
            <Link text="Read more" />
          </Badge>
        ),
        description: "Do not leave the indicator inside the element boundaries"
      }
    },
    {
      positive: {
        component: (
          <Badge>
            <Button leftIcon={WhatsNew}>What's new</Button>
          </Badge>
        ),
        description: "Choose a color that does not blend with the one of the child component"
      },
      negative: {
        component: (
          <Badge color="primary">
            <Button leftIcon={WhatsNew}>What's new</Button>
          </Badge>
        ),
        description: "Do not use a color that blends with the child component"
      }
    },
    {
      positive: {
        component: (
          <Badge alignment="circular">
            <Avatar size="medium" src={person} type="img" />
          </Badge>
        ),
        description: "When using Indicator badge, anchor it at the top-right edge"
      },
      negative: {
        component: (
          <Badge alignment="circular" anchor="top-start">
            <Avatar size="medium" src={person} type="img" />
          </Badge>
        ),
        description: "Do not place it on any other edge"
      }
    }
  ]}
/>

## Related components

<RelatedComponents componentsNames={[LINK, BUTTON, COUNTER]} />
