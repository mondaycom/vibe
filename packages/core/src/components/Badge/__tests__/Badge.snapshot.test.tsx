import { describe, it, expect } from "vitest";
import React from "react";
import renderer from "react-test-renderer";
import Badge from "../Badge";

describe("Badge", () => {
  it("renders with no props with default badge Indicator", () => {
    const tree = renderer.create(<Badge>Child</Badge>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders with Counter and counter props", () => {
    const tree = renderer
      .create(
        <Badge type="counter" count={5}>
          Child
        </Badge>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders at different position", () => {
    const tree = renderer.create(<Badge anchor="bottom-start">Child</Badge>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders with different color", () => {
    const tree = renderer.create(<Badge color="primary">Child</Badge>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders out of horizontal bounds", () => {
    const tree = renderer.create(<Badge alignment="outside">Child</Badge>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders with margins from component's bounds", () => {
    const tree = renderer.create(<Badge alignment="circular">Child</Badge>).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
