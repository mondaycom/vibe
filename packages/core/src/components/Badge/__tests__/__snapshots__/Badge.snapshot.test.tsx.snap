// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Badge > renders at different position 1`] = `
<div
  className="badgeWrapper"
>
  Child
  <div
    className="badge bottomStart rectangular"
    data-testid="badge"
  >
    <div
      className="indicator notification"
      data-testid="indicator"
    />
  </div>
</div>
`;

exports[`Badge > renders out of horizontal bounds 1`] = `
<div
  className="badgeWrapper"
>
  Child
  <div
    className="badge topEnd outside"
    data-testid="badge"
  >
    <div
      className="indicator notification"
      data-testid="indicator"
    />
  </div>
</div>
`;

exports[`Badge > renders with Counter and counter props 1`] = `
<div
  className="badgeWrapper"
>
  Child
  <div
    className="badge topEnd rectangular"
    data-testid="badge"
  >
    <span
      aria-label=" 5"
      aria-labelledby=""
      onMouseDown={[Function]}
    >
      <div
        aria-label="5"
        className="counter sizeLarge kindFill colorNegative"
      >
        <span
          data-testid="counter"
          id="counter"
        >
          5
        </span>
      </div>
    </span>
  </div>
</div>
`;

exports[`Badge > renders with different color 1`] = `
<div
  className="badgeWrapper"
>
  Child
  <div
    className="badge topEnd rectangular"
    data-testid="badge"
  >
    <div
      className="indicator primary"
      data-testid="indicator"
    />
  </div>
</div>
`;

exports[`Badge > renders with margins from component's bounds 1`] = `
<div
  className="badgeWrapper"
>
  Child
  <div
    className="badge topEnd circular"
    data-testid="badge"
  >
    <div
      className="indicator notification"
      data-testid="indicator"
    />
  </div>
</div>
`;

exports[`Badge > renders with no props with default badge Indicator 1`] = `
<div
  className="badgeWrapper"
>
  Child
  <div
    className="badge topEnd rectangular"
    data-testid="badge"
  >
    <div
      className="indicator notification"
      data-testid="indicator"
    />
  </div>
</div>
`;
